app/src/main/java/com/dreamisi/moviebase/data/Repository.kt

```diff
@@ -0,0 +1,107 @@
+package com.dreamisi.moviebase.data
+
+import com.dreamisi.moviebase.data.JsonModels.JsonGenreList
+import com.dreamisi.moviebase.data.JsonModels.JsonMovieDetails
+import com.dreamisi.moviebase.data.models.Movie
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import com.dreamisi.moviebase.data.JsonModels.JsonPopularMovies
+import com.dreamisi.moviebase.data.models.Genre
+import com.dreamisi.moviebase.data.models.MovieDetails
+import com.dreamisi.moviebase.data.services.TheMovieDataBaseAPI
+
+class Repository(private val api: TheMovieDataBaseAPI) {
+
+    private var cachedMoviesList: List<Movie>? = null
+    private var cachedGenreList: List<Genre>? = null
+    private var cachedMovieDetails: MovieDetails? = null
+
+
+    suspend fun getMoviesList(): List<Movie> = withContext(Dispatchers.IO) {
+        val cachedGenre = cachedGenreList ?: loadGenres()
+        cachedGenreList = cachedGenre
+        val cachedMovies = cachedMoviesList ?: loadMoviesFromDB(cachedGenreList ?: emptyList())
+        cachedMoviesList = cachedMovies
+        cachedMovies
+    }
```

Comment:
Давай тут с неймигом разберемся. Сейчас у тебя как-то не очень понятно в чем разница между cachedGenre (тут как будто 1 genre), cachedGenreList и пр. Плюс поправил чутка код, на самом деле можно проще без emptyList
```suggestion
    suspend fun getMoviesList(): List<Movie> = withContext(Dispatchers.IO) {
        val localGenreList = cachedGenreList ?: loadGenres()
        cachedGenreList = localGenreList
        val localMoviesList = cachedMoviesList ?: loadMoviesFromDB(localGenreList)
        cachedMoviesList = localMoviesList
        localMoviesList
    }
```